api/common/modelstatus.go

```diff
@@ -37,8 +45,40 @@ func (c *ModelStatusAPI) ModelStatus(tags ...names.ModelTag) ([]base.ModelStatus
 		return nil, err
 	}
 
-	results := make([]base.ModelStatus, len(result.Results))
-	for i, r := range result.Results {
+	if olderVersion {
+		// This caters for the fact that older version of base.ModelStatus
+		// did not have an Error field.
+		return c.processModelStatusResultsV3(result.Results)
+	}
+
+	return c.processModelStatusResults(result.Results)
+}
+
+func (c *ModelStatusAPI) processModelStatusResults(rs []params.ModelStatus) ([]base.ModelStatus, error) {
+	results := make([]base.ModelStatus, len(rs))
+	for i, r := range rs {
+		if r.Error != nil {
+			results[i].Error = errors.Trace(r.Error)
+			continue
+		}
+		model, err := names.ParseModelTag(r.ModelTag)
+		if err != nil {
+			results[i].Error = errors.Annotatef(err, "model tag %v", r.ModelTag)
+			continue
+		}
+		owner, err := names.ParseUserTag(r.OwnerTag)
+		if err != nil {
+			results[i].Error = errors.Annotatef(err, "owner tag %v", r.OwnerTag)
+			continue
+		}
+		results[i] = constructModelStatus(model, owner, r)
+	}
+	return results, nil
+}
+
+func (c *ModelStatusAPI) processModelStatusResultsV3(rs []params.ModelStatus) ([]base.ModelStatus, error) {
+	results := make([]base.ModelStatus, len(rs))
+	for i, r := range rs {
 		model, err := names.ParseModelTag(r.ModelTag)
 		if err != nil {
 			return nil, errors.Annotatef(err, "ModelTag %q at position %d", r.ModelTag, i)
```

Comment:
since the error message will contain "foo is not a valid model tag", maybe just annotate with the text "result at position %d"

Reply:
this code has been removed.